typing/ctype.ml

```diff
@@ -5475,3 +5505,26 @@ let immediacy env typ =
       else
         Type_immediacy.Always
   | _ -> Type_immediacy.Unknown
+
+let check_type_immediate env ty imm =
+  match Type_immediacy.coerce (type_immediacy_head env ty) ~as_:imm with
+  | Ok () -> Ok ()
+  | Error _ ->
+    let ty = get_unboxed_type_representation env ty in
+    Type_immediacy.coerce (type_immediacy_head env ty) ~as_:imm
```

Comment:
> Why not always get the unboxed type representation first?

You are exactly correct - this is for performance reasons.  `get_unboxed_type_representation` is potentially expensive, so in the case of a `Tconstr` we first look at the kind and see if it yields a definitive answer.  If not, we perform the expensive expansion and unboxing.

This optimization was already present in the previous version of this PR, but it was implemented by a recursive call in `check_type_immediate` in the case we needed to call `get_unboxed_type_representation`.  That approach was somewhat confusing, so here I've incorporated a suggestion from @lpw25 from that PR to eliminate the recursion by breaking it into two functions (`check_type_immediate` and `type_immediacy_head`).

The intent, then, is that `type_immediacy_head` is a cheap look at the "head" of the type, which `check_type_immediate` tries and then calls `get_unboxed_type_representation` if it doesn't provide a definitive answer.

> Why not have `kind_immediacy` handle unboxed datatypes, so that type_immediacy_head returns the correct result for all types?

`kind_immediacy` may return an approximation in two cases: unboxed types, and abbreviations (because the immediacy recorded on `Type_abstract` is a conservative upper bound, as you observe elsewhere).  As the fallback in both cases is the same (`check_type_immediate` calls `get_unboxed_type_representation`), it seemed cleanest to leave the unboxing logic there.

An alternative design would be to duplicate more of the logic of `get_unboxed_type_representation` here.  In particular, one advantage would be that we could check whether we have a definitive immediacy after each recursive step of expansion and unboxing.  But as this function and its fuel parameter are somewhat unfortunate, and it sounds like there are medium term plans to replace it, my instinct is not to do this.

So, to summarize, I do think we need some fallback in any event to deal with abbreviations, and it therefore seems cleanest not to complicate `kind_immediacy`.  But let me know if you disagree.  And in any event the situation should be documented - I'll add a comment if that sounds reasonable.

Reply:
I find the current state disturbing: there is a notion of "the immediacy of a kind", which would be an inexact approximation of "the immediacy of the type", but it is not what `kind_immediacy` currently computes. But it is not so simple to compute the "right" thing (as I suggested earlier) because it the case of unboxed datatypes we would need to recursively call `type_immediacy_head`.

In the short term, I thing maybe you could just add a comment to point out that `kind_immediacy` is incorrect on unboxed datatypes.

Reply:
I have changed the name to `kind_immediacy_approx` and added a comment to the mli.